addons/stock/models/stock_move_line.py

```diff
@@ -535,14 +535,37 @@ def _get_similar_move_lines(self):
 
     def _create_and_assign_production_lot(self):
```

Comment:
To avoid performance issue: 
```python
    def _create_and_assign_production_lot(self):
        """ Creates and assign new production lots for move lines."""
        def uniq_key_lot(ml):
            return (ml.company_id.id, ml.lot_name, ml.product_id.id)

        lot_vals = []
        # It is possible to have multiple time the same lot to create & assign, then we 
        # handle the case with 2 dictionary 
        key_to_index = {}  # key to index of the lot
        key_to_mls = defaultdict(lambda: self.env['stock.move.line'])  # key to all mls
        for ml in self:
            key = uniq_key_lot(ml)
            key_to_mls[key] |= ml 
            if key not in key_to_index:
                key_to_index[key] = len(lot_vals)
                lot_vals.append({
                    'company_id': ml.company_id.id,
                    'name': ml.lot_name,
                    'product_id': ml.product_id.id
                })

        lots = self.env['stock.production.lot'].create(lot_vals)
        for key, mls in key_to_mls.items():
            mls._assign_production_lot(lots[key_to_index[key]].with_prefetch(lots._ids))  # With prefetch to reconstruct the ones broke by accessing by index

    def _assign_production_lot(self, lot):
        self.lot_id = lot.id
```
And in product_expiry module:
```python
    def _assign_production_lot(self, lot):
        super()._assign_production_lot(lot)
        self.lot_id._update_date_values(max(self.filtered('expiration_date').expiration_date, default=False))
```